File: note_service/app/internal/events/publisher.go
package events

import (
	"context"
	"encoding/json"
	"fmt"
	"sync"
	"time"

	"github.com/google/uuid"
	amqp "github.com/rabbitmq/amqp091-go"

	"note_service/internal/config"
	handlermodel "note_service/internal/handlers/notes"
	"note_service/pkg/logging"
)

type Publisher interface {
	PublishNoteCreated(ctx context.Context, note handlermodel.Note) error
	PublishNoteUpdated(ctx context.Context, note handlermodel.Note) error
	PublishNoteDeleted(ctx context.Context, payload NoteDeletedPayload) error
	Close() error
}

type noopPublisher struct{}

func (noopPublisher) PublishNoteCreated(context.Context, handlermodel.Note) error  { return nil }
func (noopPublisher) PublishNoteUpdated(context.Context, handlermodel.Note) error  { return nil }
func (noopPublisher) PublishNoteDeleted(context.Context, NoteDeletedPayload) error { return nil }
func (noopPublisher) Close() error                                                 { return nil }

type rabbitPublisher struct {
	logger   logging.Logger
	url      string
	exchange string

	mu         sync.Mutex
	connection *amqp.Connection
	channel    *amqp.Channel
}

func NewPublisher(cfg *config.Config, logger logging.Logger) Publisher {
	if !cfg.RabbitMQ.Enabled {
		logger.Info("rabbitmq publisher disabled")
		return noopPublisher{}
	}

	publisher := &rabbitPublisher{
		logger:   logger.With("component", "rabbitmq_publisher"),
		url:      cfg.RabbitMQ.URL,
		exchange: cfg.RabbitMQ.Exchange,
	}

	if err := publisher.connect(); err != nil {
		logger.Warn("failed to initialize rabbitmq publisher, will retry on demand", "error", err)
	}

	return publisher
}

func (p *rabbitPublisher) PublishNoteUpdated(ctx context.Context, note handlermodel.Note) error {
	return p.publishNoteChange(ctx, NoteUpdatedEventType, note)
}

func (p *rabbitPublisher) PublishNoteCreated(ctx context.Context, note handlermodel.Note) error {
	return p.publishNoteChange(ctx, NoteCreatedEventType, note)
}

func (p *rabbitPublisher) PublishNoteDeleted(ctx context.Context, payload NoteDeletedPayload) error {
	if err := p.ensureConnection(); err != nil {
		return err
	}

	envelope := Envelope[NoteDeletedPayload]{
		EventID:      uuid.NewString(),
		EventType:    NoteDeletedEventType,
		EventVersion: 1,
		OccurredAt:   time.Now().Unix(),
		Producer:     "note_service",
		Payload:      payload,
	}

	body, err := json.Marshal(envelope)
	if err != nil {
		return fmt.Errorf("marshal note.deleted event: %w", err)
	}

	return p.publish(ctx, envelope.EventType, envelope.EventID, envelope.EventType, envelope.OccurredAt, body)
}

func (p *rabbitPublisher) publishNoteChange(ctx context.Context, eventType string, note handlermodel.Note) error {
	if err := p.ensureConnection(); err != nil {
		return err
	}

	envelope := Envelope[NoteUpdatedPayload]{
		EventID:      uuid.NewString(),
		EventType:    eventType,
		EventVersion: 1,
		OccurredAt:   time.Now().Unix(),
		Producer:     "note_service",
		Payload: NoteUpdatedPayload{
			NoteUUID:       note.Uuid,
			UserUUID:       note.UserUuid,
			WorkspaceID:    note.WorkspaceID,
			AuthorUserUUID: note.AuthorUserUUID,
			CategoryUUID:   note.CategoryUuid,
			Header:         note.Header,
			Body:           note.Body,
			ShortBody:      note.ShortBody,
			Tags:           note.Tags,
			Event:          note.Event,
			CreatedAt:      note.CreatedDate,
			UpdatedAt:      note.UpdatedAt,
		},
	}

	body, err := json.Marshal(envelope)
	if err != nil {
		return fmt.Errorf("marshal %s event: %w", eventType, err)
	}

	return p.publish(ctx, envelope.EventType, envelope.EventID, envelope.EventType, envelope.OccurredAt, body)
}

func (p *rabbitPublisher) publish(ctx context.Context, routingKey, eventID, eventType string, occurredAt int64, body []byte) error {
	p.mu.Lock()
	defer p.mu.Unlock()

	if p.channel == nil {
		return fmt.Errorf("rabbitmq channel is not initialized")
	}

	if publishErr := p.channel.PublishWithContext(
		ctx,
		p.exchange,
		routingKey,
		false,
		false,
		amqp.Publishing{
			ContentType:  "application/json",
			DeliveryMode: amqp.Persistent,
			MessageId:    eventID,
			Timestamp:    time.Unix(occurredAt, 0),
			Type:         eventType,
			Body:         body,
		},
	); publishErr != nil {
		return fmt.Errorf("publish %s event: %w", eventType, publishErr)
	}

	return nil
}

func (p *rabbitPublisher) Close() error {
	p.mu.Lock()
	defer p.mu.Unlock()

	var closeErr error
	if p.channel != nil {
		if err := p.channel.Close(); err != nil {
			closeErr = err
		}
		p.channel = nil
	}
	if p.connection != nil {
		if err := p.connection.Close(); err != nil && closeErr == nil {
			closeErr = err
		}
		p.connection = nil
	}

	return closeErr
}

func (p *rabbitPublisher) ensureConnection() error {
	p.mu.Lock()
	defer p.mu.Unlock()

	if p.connection != nil && !p.connection.IsClosed() && p.channel != nil {
		return nil
	}

	return p.connectLocked()
}

func (p *rabbitPublisher) connect() error {
	p.mu.Lock()
	defer p.mu.Unlock()

	return p.connectLocked()
}

func (p *rabbitPublisher) connectLocked() error {
	if p.channel != nil {
		_ = p.channel.Close()
		p.channel = nil
	}
	if p.connection != nil {
		_ = p.connection.Close()
		p.connection = nil
	}

	connection, err := amqp.Dial(p.url)
	if err != nil {
		return fmt.Errorf("dial rabbitmq: %w", err)
	}

	channel, err := connection.Channel()
	if err != nil {
		_ = connection.Close()
		return fmt.Errorf("open rabbitmq channel: %w", err)
	}

	if err = channel.ExchangeDeclare(
		p.exchange,
		"topic",
		true,
		false,
		false,
		false,
		nil,
	); err != nil {
		_ = channel.Close()
		_ = connection.Close()
		return fmt.Errorf("declare exchange %s: %w", p.exchange, err)
	}

	p.connection = connection
	p.channel = channel
	p.logger.Info("rabbitmq publisher connected", "exchange", p.exchange)
	return nil
}

File: category_service/app/messaging/publisher.py
import json
import logging
import threading
import time

import pika

from config import Config


class CategoryEventsPublisher:
    def __init__(self, config: Config, logger: logging.Logger) -> None:
        self.config = config
        self.logger = logger
        self._lock = threading.Lock()
        self._connection = None
        self._channel = None

    def publish_category_updated(
        self,
        *,
        category_uuid: str,
        workspace_id: str,
        actor_user_uuid: str | None,
        name: str | None,
    ) -> None:
        if not self.config.RABBITMQ_ENABLED or not name:
            return

        envelope = {
            "event_id": f"category-{category_uuid}-{int(time.time() * 1000)}",
            "event_type": "category.updated",
            "event_version": 1,
            "occurred_at": int(time.time()),
            "producer": "category_service",
            "payload": {
                "category_uuid": category_uuid,
                "workspace_id": workspace_id,
                "actor_user_uuid": actor_user_uuid or "",
                "name": name,
            },
        }
        self._publish("category.updated", envelope)

    def close(self) -> None:
        with self._lock:
            if self._channel and self._channel.is_open:
                try:
                    self._channel.close()
                except Exception:
                    self.logger.debug("failed to close category publisher channel", exc_info=True)
            if self._connection and self._connection.is_open:
                try:
                    self._connection.close()
                except Exception:
                    self.logger.debug("failed to close category publisher connection", exc_info=True)
            self._channel = None
            self._connection = None

    def _publish(self, routing_key: str, payload: dict) -> None:
        self._ensure_connection()
        body = json.dumps(payload).encode("utf-8")

        with self._lock:
            self._channel.basic_publish(
                exchange=self.config.RABBITMQ_EXCHANGE,
                routing_key=routing_key,
                body=body,
                properties=pika.BasicProperties(
                    content_type="application/json",
                    delivery_mode=2,
                    message_id=payload["event_id"],
                    type=payload["event_type"],
                    timestamp=payload["occurred_at"],
                ),
            )

    def _ensure_connection(self) -> None:
        with self._lock:
            if self._connection and self._connection.is_open and self._channel and self._channel.is_open:
                return

            if self._channel and self._channel.is_open:
                self._channel.close()
            if self._connection and self._connection.is_open:
                self._connection.close()

            self._connection = pika.BlockingConnection(pika.URLParameters(self.config.RABBITMQ_URL))
            self._channel = self._connection.channel()
            self._channel.exchange_declare(
                exchange=self.config.RABBITMQ_EXCHANGE,
                exchange_type="topic",
                durable=True,
            )
            self.logger.info("category events publisher connected")
